src/server/ttl/ttl_checker.rs

```diff
@@ -0,0 +1,276 @@
+                    error!("ttl checker: failed to get next region information";
+                        "err" => ?e);
+                    TTL_CHECKER_ACTIONS_COUNTER_VEC
+                        .with_label_values(&["error"])
+                        .inc();
+                    continue;
+                }
+            }
+            break;
+        }
+        TTL_CHECKER_ACTIONS_COUNTER_VEC
+            .with_label_values(&["finish"])
+            .inc();
+        info!(
+            "ttl checker finishes a round, wait {}s to start next round",
+            self.poll_interval.as_secs()
+        );
+        // make sure the data point of metrics is pulled
+        thread::sleep(Duration::from_secs(40));
+        TTL_CHECKER_PROCESSED_REGIONS_GAUGE.set(0);
+    }
+
+    fn get_interval(&self) -> Duration {
+        self.poll_interval
+    }
+}
+
+fn check_ttl_for_range<E: KvEngine>(
+    engine: &E,
+    start_key: &[u8],
+    end_key: &[u8],
+    without_l0: bool,
+) {
+    let current_ts = UnixSecs::now().into_inner();
+    let mut files = Vec::new();
+    let res = match engine.get_range_ttl_properties_cf(CF_DEFAULT, start_key, end_key) {
+        Ok(v) => v,
+        Err(e) => {
+            error!(
+                "get range ttl properties failed";
+                "range_start" => log_wrappers::Value::key(&start_key),
+                "range_end" => log_wrappers::Value::key(&end_key),
+                "err" => %e,
+            );
+            TTL_CHECKER_ACTIONS_COUNTER_VEC
+                .with_label_values(&["error"])
+                .inc();
+            return;
+        }
+    };
+    if res.is_empty() {
+        TTL_CHECKER_ACTIONS_COUNTER_VEC
+            .with_label_values(&["empty"])
+            .inc();
+        return;
+    }
+    for (file_name, prop) in res {
+        if prop.max_expire_ts <= current_ts {
+            files.push(file_name);
+        }
+    }
+    if files.is_empty() {
+        TTL_CHECKER_ACTIONS_COUNTER_VEC
+            .with_label_values(&["skip"])
+            .inc();
+        return;
+    }
+
+    let timer = Instant::now();
+    let files_count = files.len();
+    for file in files {
+        let compact_range_timer = TTL_CHECKER_COMPACT_DURATION_HISTOGRAM.start_coarse_timer();
+        if let Err(e) = engine.compact_files_cf(CF_DEFAULT, vec![file], None, 0, without_l0) {
+            error!(
+                "execute ttl compact files failed";
+                "range_start" => log_wrappers::Value::key(&start_key),
+                "range_end" => log_wrappers::Value::key(&end_key),
+                "err" => %e,
+            );
+            TTL_CHECKER_ACTIONS_COUNTER_VEC
+                .with_label_values(&["error"])
+                .inc();
+            continue;
+        }
+        compact_range_timer.observe_duration();
+        TTL_CHECKER_ACTIONS_COUNTER_VEC
+            .with_label_values(&["compact"])
+            .inc();
+        thread::sleep(Duration::from_secs(2));
+    }
+
+    info!(
```

Comment:
The number of log  record may be too large when there are too many region need to process every round